python/src/opendp/mod.py

```diff
@@ -233,85 +318,120 @@ def to_str(self, data):
-            ptr, len_ = array, len(val)
-        else:
-            raise Exception("Unknown type", type(val))
-        return ctypes.byref(FfiSlice(ctypes.cast(ptr, ctypes.c_void_p), len_))
-
-    def py_to_obj(self, val):
-        type_name = self.get_ffi_type_name(val)
-        type_args = f"<{type_name}>".encode()
-        raw = self.to_raw(val)
-        return self.data.object_new(type_args, raw)
-
-    def from_raw(self, type_name, raw):
-        if type_name == "i32":
-            return ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_int32)).contents.value
-        elif type_name == "u32":
-            return ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_uint32)).contents.value
-        elif type_name == "f64":
-            return ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_double)).contents.value
-        elif type_name == "String":
-            return ctypes.cast(raw.contents.ptr, ctypes.c_char_p).value.decode()
-        elif type_name.startswith("Vec<"):
-            if type_name == "Vec<i32>":
-                array = ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_int32))
-            elif type_name == "alloc::vec::Vec<f64>":
-                array = ctypes.cast(raw.contents.ptr, ctypes.POINTER(ctypes.c_double))
-            else:
-                raise Exception("Unknown type", type_name)
-            return array[0:raw.contents.len]
+        raise Exception("Unknown type", type(val))
+
+    def to_raw(self, val, type_name):
+        if type_name in self.ATOM_MAP:
+            return scalar_to_raw(val, type_name)
+
+        if type_name.startswith("Vec<") and type_name.endswith('>'):
+            return vector_to_raw(val, type_name)
+
+        if type_name.startswith('(') and type_name.endswith(')'):
+            return tuple_to_raw(val, type_name)
+
+        if type_name == "String":
+            return string_to_raw(val)
+
+        raise Exception("Unknown type name", type_name)
+
+    def py_to_obj(self, val, type_arg=None):
+        if type_arg:
+            type_name = type_arg[1:-1]
         else:
-            raise Exception("Unknown type", type_name)
+            type_name = self.infer_ffi_type_name(val)
+
+        raw = self.to_raw(val, type_name)
+        return self.data.object_new(f"<{type_name}>".encode(), raw)
+
+    def from_raw(self, raw, type_name):
+        if type_name in self.ATOM_MAP:
+            return raw_to_scalar(raw, type_name)
+
+        if type_name.startswith("Vec<") and type_name.endswith('>'):
+            return raw_to_vector(raw, type_name)
+
+        if type_name.startswith('(') and type_name.endswith(')'):
+            return raw_to_tuple(raw, type_name)
+
+        if type_name == "String":
+            return raw_to_string(raw)
+
+        raise Exception("Unknown type name", type_name)
 
     def obj_to_py(self, obj):
         type_name = self.data.object_type(obj).value.decode()
         raw = self.data.object_as_raw(obj)
```

Comment:
There's a leak here. The `FfiSlice` in raw is heap allocated, should be freed with a call to `opendp_data__slice_free()`. (This frees just the slice struct itself, not the inner pointer, which is handled when the `FfiObject` is freed.

Sorry, this was my omission when I first made these changes.

Reply:
In object_new, `FfiObject::new` uses the `FfiOwnership::LIBRARY` variant. I think that's also incorrect- in this case the memory is owned by python. So when we clean up the FfiObject in `from_raw`, ManuallyDrop will be triggered and it will drop the python-owned memory.

Reply:
I take it back, we're not dropping FfiObject, we're dropping FfiSlice, as you say, so it won't free anything unexpected.